### PR TITLE
Optimized ScriptEditor initialization when many scripts are loaded

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
This change avoids the editor to freeze for several seconds when a project with lots of open scripts (> 50) is loaded in the editor.

It focuses on a few heavy operations previously executed on all loaded scripts:
- Initialize script resource (script validation/parsing) now only on focus
- ScriptTextEditor: code editor and menus are added to the scene only on focus
- Add to "recently opened" now only when opening a new script (causes load/save metadata file)

*Results*
Loading time 5x faster with ~100 scripts in release_debug
(1s instead of 5s on my configuration)

*Tests*
I've added safety nets in different places to avoid potential crashes, and tested as much as I could with gdscript, visual scripts and text files, but it probably needs more intensive tests to catch any potential remaining issue.

### CODE REVIEW

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Haven't given it a proper look over, but delaying the script loading will break the state restore on load e.g breakpoints, bookmarks etc. 

Have you got any rough benchmarks for each change?  At a guess just delaying any calls to `_validate_script` should see similar results.

**Comment by pouleyKetchoupp** [comment]
@Paulb23 Ok, I missed the fact bookmarks and breakpoints are restored on project load and I haven't tested that at all. I'm going to check it and make sure I didn't break this functionality.

And I'll post some proper benchmarks, I don't have all the measurements at the moment.

In the meantime, here's some details about how I've chosen those different areas of improvement:
These changes are based on delaying all the most expensive operations I've measured using visual studio profiler. From what I remember, `_validate_script` was about a half of what I've saved. Another large portion is about adding all the controls (calls to `add_child`) when initializing the text editor and menus for each tab. Also, adding all the scripts to the "recently opened" menu takes a lot of time because it opens and saves a metadata file on each iteration, and it doesn't seem relevant when we're restoring previous script tabs anyway.

**Comment by pouleyKetchoupp** [comment]
@Paulb23  I've pushed some changes that fix restoring the state for breakpoints, bookmarks, folded lines and selection.

I've also made some more detailed performance measurements on a project with 100 scripts.
(for some reason it takes 3s without optimization instead of the 5s I got before, but the performance improvement is still around 80%).

**Loading time with no optimization: ~3000ms**
*Profiling details:*
set_edited_resource: 650ms
_add_recent_script: 600ms
add_child (tab): 550ms
create_editor: 350ms
_go_to_tab: 300ms
load_resource: 100ms
other: 450ms

**Loading time with full optimization: ~500ms**
*Saved time approximation for each change:*
Disabled tab switch for each loaded script: saves ~300ms
Disabled add to recent files for each loaded script: saves ~700ms
Delayed set_edited_resource: saves ~700ms
Delayed add/initialize code editor controls: saves ~800ms

**Comment by Paulb23** [comment]
@pouleyKetchoupp thanks! those benchmarks are surprising.

Although I've just looked at `_add_recent_script` and does indeed seems a little excessive in file I/O. Rather then delaying the call, have it write out once on close and read once on engine start, otherwise use the `popup` menu / vector.  If it needs to be kept up to date then it could write on script open, except for launch when `resorting_layout` is true. But indeed, no need to read, write then read what just been written.

With regards to `set_edited_resource`, how much of it is loading the `Script` and `TextEdit` etc vs `_validate_script`? I'm very sceptical of adding lazy loading here due to the large amount of complexity that it brings, unless there's a significant benefit / we can find a clean interface for it (#28607).

One such case, if you set a breakpoint, restart the editor, then without opening the script, hit play, does the breakpoint still trigger?

**Comment by pouleyKetchoupp** [comment]
@Paulb23 Thanks for your feedback.

About `_add_recent_script`:
What you propose makes sense as an optimization, but in the case of restoring all the script tabs from the previous sessions, why do we want them all added to recently opened scripts? I would rather see in this menu the actual last scripts I've opened in previous sessions, rather than the last few scripts that were opened while loading the project.

About `set_edited_resource`:
Here's how the time is spent in this method:
60% in `_validate_script()`
30% in `_set_theme_for_script()`
10% in `TextEdit::set_text()`
If it makes sense to you, I can try to keep the lazy initialization for the code editor itself and the theme, while still calling `set_edited_resource` in any case, just to store the script resource, validate it, and set the text in the code editor (some features like breakpoints need to parse it).
(I'm guessing here that `_validate_script` is required before doing some things like activating breakpoints and possibly some other functionalities that affect all scripts in the list).
If that works, we might end up with a performance improvement of 60-70% instead of 80%, with lighter code changes and less risk to break features.

Breakpoints triggering when the code editor is not initialized:
Good point, the case you describe is not working anymore with my current changes. I think I can find a way to fix it easily, but if we go with the solution I proposed in the previous paragraph I probably won't have to do anything special to handle that.

**Comment by Paulb23** [comment]
Taking of the optimisation hat, I don't use recent files all that much, I know QTCreator adds the scripts on launch, so have no strong feelings as to its behaviour in this regard.

As for `set_edited_resource`, thanks for benchmarking again.  I was worried that `_validate_script` would be consuming less time.

Given this, that approach would be preferred. Though it should also be safe to delay `_validate_script` until tab click. So just loading the `Script` and storing it in `TextEdit`. Will have to double check, but I think the only thing outside the script editor dependent on `_validate_script` is the connection dock, not too sure how interlinked this is or if a workaround is possible.

The breakpoint system is a bit convoluted, a signal is fired in `TextEdit` when a breakpoint is placed that calls back to `script_text_editor` which enters it into `ScriptEditorDebugger`. I don't know if / how many more systems rely on this kind of process.

**Comment by pouleyKetchoupp** [comment]
Thanks for all the details about `_validate_script` and how breakpoints work, it helps a lot. I'll try and implement this solution in the next few days.

Concerning the recent files, I'm still not convinced changing the implementation to preserve the current behavior is the way to go.
> 
> 
> Taking of the optimisation hat, I don't use recent files all that much, I know QTCreator adds the scripts on launch, so have no strong feelings as to its behaviour in this regard.

I don't know QTCreator at all, but I would like to think in term of user workflow in the context of Godot. In what use case would the user need to see the last restored files in this list? In my experience, with the current behavior, recent files are not really useful because I mostly see some help files there, just because they are at the end of the tabs.

**Comment by Paulb23** [comment]
I might have been a bit unclear, but I have no issues with changing the behaviour of recent files if it improves the workflow. Can always reduce the amount of disk usage at a later date :)

**Comment by pouleyKetchoupp** [comment]
Oh great! Glad we're on the same page then :)

**Comment by pouleyKetchoupp** [comment]
@Paulb23 I've just pushed the changes we discussed about. I haven't encountered any critical issue and everything seems to work well so far.

The result is a loading time of ~800ms in a similar test case as before, so it's a pretty close improvement!

**Comment by pouleyKetchoupp** [comment]
I've reverted some cosmetic changes to make diffs more readable.

File: editor/code_editor.cpp
Comment (Paulb23) [inline editor/code_editor.cpp]: This will need testing with folded lines and function list, as in `_members_overview_selected` we reset the scroll. Don't know if this will break the focus.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Looking good so far, just a couple of notes on the diff:

I assume all the:
```
RES resource = se->get_edited_resource();
ERR_CONTINUE(resource.is_null());
```
in `script_editor_plugin` similarly for `ERR_FAIL_COND(script.is_null());` in `script_text_editor` and `ERR_FAIL_COND(p_res.is_null());` in `text_editor`, are remnants from lazy loading, now as its always loaded, these can be removed?

Rather then exposing the tabs loaded state via `enable_editor`, it should be okay to perform the setup in the `ensure_focus` method.

**Comment by pouleyKetchoupp** [comment]
@Paulb23 I've just pushed a new version with changes from your comments.

- Checks on null edited resource:
All removed now, it was indeed not needed anymore.

- Using `ensure_focus` instead of a new method `enable_editor`:
I actually had to keep it this way, because there are several cases in the plugin where we're not setting the focus, but we do need to enable the controls for some functionalities to work.

- Grab focus on `set_edit_state`:
You're right, it did cause an issue with losing focus when selecting functions from the function list. I've fixed it by handling that directly in `ScriptTextEditor::set_edit_state`:
```
if (editor_enabled) {
    ensure_focus();
}
```

**Comment by Xrayez** [comment]
I'm currently experiencing an issue as described in #39841 which this PR can likely fix. I could try and test this PR, but would be nice to resolve merge conflicts, and probably see how this can be ported to 3.2.

**Comment by pouleyKetchoupp** [comment]
Ok, I'll rebase tomorrow so you can test your use case. Hopefully we'll be able to move forward after that because this PR has been stalled for a while :)

It should be easy to port back to 3.2 branch too, I don't think there's much difference.

**Comment by Xrayez** [comment]
Well I tested this PR in its current state already!

What took 8 minutes now takes 2-3 minutes to startup the editor for 500 opened scripts in `target=debug` (see #39841), so it doesn't completely eliminate the performance issue but definitely an improvement.

`File` → `Close All` takes similar duration to close all scripts, and received this message continously:
```
Object was deleted while awaiting a callback
```

But that's not an issue caused by this PR, but rather #35653. Yet increasing the `MessageQueue` size in project settings does not seem to help this (and works alright in 3.2, again I don't know whether that's related).

So, I'm not sure if that's inherent problem of the existing parsing/loading system (https://github.com/godotengine/godot/issues/27333#issuecomment-575184595) or whether there might be some other steps to be done to eliminate this completely, but yeah this already improves performance. `project_metadata.cfg` read/write is done at most ~3 times at project startup now.

**Comment by Xrayez** [comment]
The startup performance issue can be eliminated completely simply by not letting the script list to grow without limits.

I personally think it would be nice to add an editor setting to automatically clear the list upon exit, as I don't see much point seeing the history of previously edited scripts in other sessions. Setting a reasonable maximum list size is also a solution, at most 7-10 scripts, else doesn't make much sense to scroll down and search because that can be easily done in the `FileSystem` dock instead.

I've actually [created a plugin](https://github.com/Xrayez/godot-editor-plugin-tools/blob/master/addons/script_editor/script_editor_plugin.gd?rgh-link-date=2020-06-25T23%3A08%3A16Z) for this.

**Comment by pouleyKetchoupp** [comment]
Rebase done. Once it's approved I'll check if it needs a specific PR for 3.2.

@Xrayez Yes it would be useful to add editor options for the number of loaded scripts (though I would rather have a default value of at least 50) and for enabling/disabling the loading of previous session (enabled by default).

**Review by Paulb23 — COMMENTED** [review]
Apologies for not re-reviewing earlier, overall changes looks good.

One minor comment made inline, will need wider testing to catch any edge cases. Not sure if it's worth applying the same changes to `TextEditor` as it's very similar to `ScriptTextEditor`?

File: editor/plugins/script_text_editor.cpp
Comment (pouleyKetchoupp) [reply editor/plugins/script_text_editor.cpp:475]: Makes sense, I've just added it to `TextEditor` to keep things as safe as possible.

**Comment by pouleyKetchoupp** [comment]
Thanks @Paulb23 for reviewing again!

Concerning wider tests, unless there's a specific test plan I can go through I think the best course of action now would be to merge to master so we can start catching potential edge cases.

It makes sense to port this optimization to 3.2 too, but probably after 3.2.3 since it is a bug fix release, so we'll have more time to test on that branch as well.

@akien-mga Does that plan make sense to you?

**Comment by pouleyKetchoupp** [comment]
Just pushed again for fixes due to #40253.

**Review by Paulb23 — APPROVED** [review]

**Comment by akien-mga** [comment]
Needs a rebase, otherwise should be good to merge.

**Comment by pouleyKetchoupp** [comment]
Rebase done.
@akien-mga: I'll make a separate PR for 3.2 because the base code is too different to merge easily.

@Paulb23: After the update in syntax highlighting, I had to make an extra change because the syntax highlighting cache update was done on startup for each script and it spends a very long time here in case there are many global classes:
https://github.com/godotengine/godot/blob/9e34ba48556b71f1f672b5be47d12b0e3f679b86/modules/gdscript/editor/gdscript_highlighter.cpp#L447-L453

So I've moved `_load_theme_settings` call from ready notification to `enable_editor` in both `TextEditor` and `ScriptTextEditor` and it seems to work fine.

File: editor/plugins/script_text_editor.cpp
Comment (Xrayez) [inline editor/plugins/script_text_editor.cpp:1788]: CI complains here: 

```
editor/plugins/script_text_editor.cpp: In constructor 'ScriptTextEditor::ScriptTextEditor()':
editor/plugins/script_text_editor.cpp:1788:34: error: ambiguous overload for 'operator=' (operand types are 'Ref<EditorSyntaxHighlighter>' and 'std::nullptr_t')
 1788 |  highlighters[TTR("Standard")] = nullptr;
      |                                  ^~~~~~~
```

```suggestion
	highlighters[TTR("Standard")] = Ref<EditorSyntaxHighlighter>();
```

Comment (Paulb23) [reply editor/plugins/script_text_editor.cpp:1788]: The line can be removed, the "Standard" one is no longer `null` and they are both already generated and added [here.](https://github.com/godotengine/godot/blob/ca1a5bdcbe28c38b745a0c0cbdbbe42708069e35/editor/plugins/script_text_editor.cpp#L1719-L1726)

File: editor/plugins/script_text_editor.cpp
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1790]: Same here `Standard` is already added above. In addition it will have already selected the highlighter, so the selected item will probably need re-selecting.

Comment (pouleyKetchoupp) [reply editor/plugins/script_text_editor.cpp]: Thanks! I had missed some changes in the constructor from your PR. I've just removed this line and also moved the addition of the plain and standard highlighters back to the constructor, so the order of items and the selection are not messed up anymore.

**Comment by pouleyKetchoupp** [comment]
Changes should be ready to merge with the last force push.

@Paulb23 Please confirm or let me know if there's anything else I should re-test after the highlighter changes (I've checked highlighting in script & plain text, switching type in highlighter menu and modified highlighter reloaded successfully when restarting the editor).

**Comment by Paulb23** [comment]
Tested it locally, looks good. One issue with the highlighters, when opening a `.gd` file the correct highlighter is used, but the drop down shows that both `GDScript` and `Standard` are selected.

**Comment by pouleyKetchoupp** [comment]
@Paulb23 I've just changed `set_syntax_highlighter` in `TextEditor` and `ScriptTextEditor` to deactivate other radio buttons instead of doing it only in the internal `_change_syntax_highlighter`. Now it should be safer and the issue on opened scripts is solved.

**Review by Paulb23 — APPROVED** [review]
Yeah that's a lot cleaner. LGTM.

**Review by Xrayez — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
